pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -316,6 +316,242 @@ func (p *staticPolicy) updateCPUsToReuse(pod *v1.Pod, container *v1.Container, c
+	}
+
+	return nil
+}
+
+// This function is the entry point for pod-level resource allocation.
+// It's called once per pod by the Topology Manager's pod-scope admit handler.
+// The logic here allocates a single "bubble" of CPUs for the entire pod
+// and then partitions that bubble among the containers.
+func (p *staticPolicy) AllocatePod(logger logr.Logger, s state.State, pod *v1.Pod, hint topologymanager.TopologyHint) error {
+	// Centralized check: Do nothing if this pod is not managed at the pod level.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResourceManagers) || !resourcehelper.IsPodLevelResourcesSet(pod) {
+		return nil
+	}
+
+	// 1. Validate for the "empty shared pool" case.
+	if err := p.validatePodScopeResources(logger, pod); err != nil {
+		return &admission.EmptyPodSharedPoolError{Msg: err.Error()}
+	}
+
+	logger.V(4).Info("AllocatePod called for pod-level managed pod", "pod", klog.KObj(pod), "hint", hint.NUMANodeAffinity)
+	podUID := string(pod.UID)
+
+	// 2. Calculate the total number of CPUs required for the pod, considering init container reuse.
+	totalPodCPUs := p.podGuaranteedCPUs(logger, pod)
+	logger.V(4).Info("Calculated total pod CPUs", "pod", klog.KObj(pod), "totalCPUs", totalPodCPUs)
+
+	// 3. Enforce SMT alignment policy if FullPhysicalCPUsOnly is enabled.
+	if err := p.enforceSMTAlignment(s, totalPodCPUs); err != nil {
+		return err
+	}
+
+	// 4. Allocate the entire CPU "bubble" for the pod using the hint from the Topology Manager.
+	podAllocation, err := p.allocateCPUs(logger, s, totalPodCPUs, hint.NUMANodeAffinity, cpuset.New())
+	if err != nil {
+		logger.Error(err, "Unable to allocate CPUs for pod", "totalPodCPUs", totalPodCPUs)
+		return err
+	}
+	logger.V(4).Info("Allocated pod-level CPU bubble", "pod", klog.KObj(pod), "allocation", podAllocation.CPUs.String())
+
+	// 5. Partition the pod's allocation, handling init container CPU reuse correctly.
+	reusableCpus := cpuset.New()
+	exclusiveCPUs := make(map[string]cpuset.CPUSet)
+	podSharedPool := podAllocation.CPUs.Clone()
+
+	// First, iterate through init containers, allocating CPUs and identifying reusable sets.
+	for _, c := range pod.Spec.InitContainers {
+		if numCPUs := p.guaranteedCPUs(logger, pod, &c); numCPUs > 0 {
+			metrics.CPUManagerPinningRequestsTotal.Inc()
+			runnable := reusableCpus.Union(podSharedPool)
+			cset, err := p.takeByTopology(logger, runnable, numCPUs)
+			if err != nil {
+				metrics.CPUManagerPinningErrorsTotal.Inc()
+				metrics.ResourceManagerAllocationErrorsTotal.WithLabelValues("cpu", "pod").Inc()
+				return err
+			}
+			metrics.ResourceManagerAllocationsTotal.WithLabelValues("cpu", "pod").Inc()
+			exclusiveCPUs[c.Name] = cset
+
+			if !podutil.IsRestartableInitContainer(&c) {
+				reusableCpus = reusableCpus.Union(cset)
+			}
+			podSharedPool = podSharedPool.Difference(cset)
+		}
+	}
+
+	// Second, iterate through regular containers, allocating from the remaining pool.
```

Comment:
When we reach this point, we have the full cpu "bubble" available again, because all init container terminated.
It would be nice to make this explicitly recorded in the code, perhaps recomputing the pools from scratch again to avoid carrying around accounting errors. WDYT?